web_app/order_books/haiko/main.py

```diff
@@ -0,0 +1,246 @@
+            raise RuntimeError(f"Unexpected error from API: {supported_tokens}")
+        supported_tokens_filtered = [
+            token
+            for token in supported_tokens
+            if token["address"] in (self.token_a, self.token_b)
+        ]
+        if len(supported_tokens_filtered) != 2:
+            raise ValueError("One of tokens isn't supported by Haiko")
+
+    def set_current_price(self, current_price: Decimal) -> None:
+        """
+        Set the current price based on the current tick.
+        :param current_price: Decimal - Current market price
+        """
+        self.current_price = current_price
+
+    def sort_asks_bids(self) -> None:
+        """Sort bids and asks data in correct order."""
+        self.asks.sort(key=lambda ask: ask[0])
+        self.bids.sort(key=lambda bid: bid[0], reverse=True)
+
+    def fetch_price_and_liquidity(self) -> None:
+        tokens_markets = self._filter_markets_data(
+            self.haiko_connector.get_pair_markets(self.token_a, self.token_b)
+        )
+        latest_block_info = self.blast_connector.get_block_info()
+        if latest_block_info.get("error"):
+            raise RuntimeError(f"Blast-api returned an error: {latest_block_info}")
+
+        self.block = latest_block_info["result"]["block_number"]
+        self.timestamp = latest_block_info["result"]["timestamp"]
+
+        if not tokens_markets:
+            message = "Markets for this pair isn't available for now"
+            self.logger.critical(f"Pair of tokens: {self.token_a}-{self.token_b}")
+            self.logger.critical(f"{message}\n")
+            return
+
+        for market in tokens_markets:
+            market_id = market["marketId"]
+            market_depth_list = self.haiko_connector.get_market_depth(market_id)
+            if not market_depth_list:
+                self.logger.info(f"Market depth for market {market_id} is empty.")
+                continue
+            self._calculate_order_book(market_depth_list, Decimal(market["currPrice"]))
+
+        self.sort_asks_bids()
+        self.current_price = max(tokens_markets, key=lambda x: Decimal(x["tvl"]))["currPrice"]
+
+    def _calculate_order_book(
+        self, market_ticks_liquidity: list, current_price: Decimal
+    ) -> None:
+        self.set_current_price(current_price)
+        price_range = self.calculate_price_range()
+        asks, bids = [], []
+        for tick_info in market_ticks_liquidity:
+            tick_info["price"] = Decimal(tick_info["price"])
+            tick_info["liquidityCumulative"] = Decimal(tick_info["liquidityCumulative"])
+            if tick_info["price"] >= current_price:
+                asks.append(tick_info)
+            else:
+                bids.append(tick_info)
+
+        if not asks or not bids:
+            return
+
+        bids.sort(key=lambda x: x["price"], reverse=True)
+        asks.sort(key=lambda x: x["price"])
+        self.add_bids(bids, price_range)
```

Comment:
Bids quantities are currently denominated in quote tokens. However we need all the order book to be in base tokens.